eng/scripts/Generate-ServiceDirectories-From-Project-List.ps1

```diff
@@ -0,0 +1,116 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Use case:
+# Given a project list, set following environment variables in JSON format:
+# 1. SparseCheckoutDirectories - This the list of sparse checkout paths that will be used by sparse-checkout.yml
+# 2. ServiceDirectories - A list of ServiceDirectories.
+#
+# The project list should be generated by a call to generate-project-list-and-cache-maven-repository.yml
+# prior to calling this script.
+param(
+  # $(Build.SourcesDirectory) - root of the repository
+  [Parameter(Mandatory=$true)][string]$SourcesDirectory,
+  # ArtifactsList will be using ('${{ convertToJson(parameters.Artifacts) }}' | ConvertFrom-Json | Select-Object name, groupId)
+  [Parameter(Mandatory=$true)][array] $ProjectList
+)
+
+$StartTime = $(get-date)
+. "${PSScriptRoot}/../common/scripts/common.ps1"
+$Path = Resolve-Path ($PSScriptRoot + "/../../")
+$script:FoundError = $false
+
+function Build-Unreleased-List-From-File {
+    param([string]$versionFile)
+    $unreleasedList = @()
+    foreach($line in Get-Content $versionFile) {
+        if (!$line -or $line.Trim() -eq '' -or $line.StartsWith("#") -or !$line.StartsWith("unreleased_")) {
+            continue
+        } else {
+            $split = $line.Split(";")
+            if ($split.Length -ne 2) {
+                LogError("Malformed unreleased_ dependency line in $versionFile. line=$line")
+                # If there's a malformed unreleased_ dependency line in version_client.txt then
+                # immediately exit. This is the only place we need to immediately exit.
+                exit(1)
+            }
+            $temp = $split[0].Replace("unreleased_","")
+            $unreleasedList += $temp
+        }
+    }
+    return ,$unreleasedList
+}
+
+$unreleasedList = @()
+# version_client is the only version file that will have unreleased_ dependencies as these
+# are disallowed for data track libraries.
+$unreleasedList = Build-Unreleased-List-From-File $Path\eng\versioning\version_client.txt
+
+$sparseCheckoutDirHash = @{}
+$serviceDirHash = @{}
+foreach($file in Get-ChildItem -Path $SourcesDirectory -Filter pom*.xml -Recurse -Depth 3 -File) {
```

Comment:
You should consider using Get-java-PackageInfoFromRepo from language-settings.ps1 to try and start using a common function for gathering properties for the projects.